sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+    
+    auto AssertBufferAcc = MQueue->get_context()->getAssertBufferAccessor(this);
+
+    parallel_for_impl<KName2>(
+      Range,
+      [=](Item) {
+        __store_acc(AssertBuffAcc);
+        KernelFunc(Item);
+      });
+  } else {
+#endif
+
+     // (No assert required) OR (Assert supported by device)
+     // ordinary enqueue process
+
+#ifndef NDEBUG
+  }
+#endif
+}
+
+}
+```
+
+
+#### Built-ins operation
+
+Accessor is a pointer augmented with offset and two ranges (access range and
+memory range).
+
+There are two built-ins provided by frontend:
+ * `__store_acc()` - to store accessor metadata into program-scope variable.
+ * `__store_assert_failure()` - to store flag about assert failure in a buffer
+   using the metadata stored in program-scope variable.
+
+The accessor should be stored to program scope variable in global address space
+using atomic operations. Motivation for using atomic operations: the program may
+contain several kernels and some of them could be running simultaneously on a
+single device.
+
+The `__store_assert_failure()` built-in atomically sets a flag in a buffer. The
+buffer is accessed using accessor metadata from program-scope variable. This
+built-in return a boolean value which is `true` if the flag is set by this call
+to `__store_assert_failure()` and `false` if the flag was already set.
+Motivation for using atomic operation is the same as with `__store_acc()`
+builtin.
+
+The following pseudo-code snippets shows how these built-ins are used.
+First of all, assume the following code as user's one:
+```
+void user_func(int X) {
+  assert(X && “X is nil”);
+}
+
+int main() {
+  queue Q(...);
+  Q.submit([&] (handler& CGH) {
+    CGH.single_task([=] () {
+      do_smth();
+      user_func(0);
+      do_smth_else();
+    });
+  });
+  ...
+}
+```
+
+The following LLVM IR pseudo code will be generated for the user's code:
+```
+@AssertBufferPtr = global void* null
+@AssertBufferAccessRange = ...
+@AssertBufferMemoryRange = ...
+@AssertBufferOffset = ...
+
+/// user's code
+void user_func(int X) {
+if (!(X && “X is nil")) {
+    __assert_fail(...);
+  }
+}
+
+users_kernel(...) {
+  do_smth()
+  user_func(0);
+  do_smth_else();
+}
+
+/// a wrapped user's kernel
+kernel(AssertBufferAccessor, OtherArguments...) {
+  __store_acc(AssertBufferAccessor);
+  users_kernel(OtherArguments...);
+}
+
+/// __assert_fail belongs to Linux version of devicelib
+void __assert_fail(...) {
+  ...
+  __devicelib_assert_fail(...);
+}
+
+void __devicelib_assert_fail(Expr, File, Line, GlobalID, LocalID) {
+  ...
+  if (__store_assert_info())
+    printf("Assertion `%s' failed in %s at line %i. GlobalID: %i, LocalID: %i",
+           Expr, File, Line, GlobalID, LocalID);
+}
+
+/// The following are built-ins provided by frontend
+void __store_acc(accessor) {
+  %1 = accessor.getPtr();
+  store void * %1, void * @AssertBufferPtr
+}
+
+bool __store_assert_info(...) {
+  AssertBAcc = __fetch_acc();
```

Comment:
This built-in is not mentioned above. I also think that it might be easier for FE to just provide a single built-in, which returns an address of a global variable - the rest of operations could be done in DPC++ Headers or within the assert fallback implementation, including the atomic update of that buffer.

Moreover, if you introduce a way for SYCL kernel function to access non-const global variable (for example only ones which are marked with special attribute) then you can implement all built-ins entirely in DPC++ Headers

Reply:
> Moreover, if you introduce a way for SYCL kernel function to access non-const global variable (for example only ones which are marked with special attribute) then you can implement all built-ins entirely in DPC++ Headers

This one sounds good. 

@kbsmith-intel , @GarveyJoe , @kbobrovs , what do you think?

Reply:
Documented approach with program-scope variable.